homeassistant/components/sensor/__init__.py

```diff
@@ -476,6 +481,23 @@ def state(self) -> Any:
                     f"Sensor {self.entity_id} provides state value '{value}', "
                     "which is not in the list of options provided"
                 )
+            return value
+
+        # If the sensor has neither a device class, a state class nor
+        # a unit_of measurement then there are no further checks or conversions
+        if not device_class and not state_class and not unit_of_measurement:
+            return value
+
+        if not isinstance(value, (int, float, Decimal)):
+            try:
+                _ = float(value)  # type: ignore[arg-type]
+            except ValueError as err:
+                raise ValueError(
+                    f"Sensor {self.entity_id} has device class {device_class}, "
+                    f"state class {self.state_class} and unit {unit_of_measurement} "
+                    "thus indicating it has a numeric value; however, it has the "
+                    f"non-numeric value: {value} ({type(value)})"
+                ) from err
```

Comment:
Just wondering... is there an appetite for this check?
Or maybe we should only log and return existing value for a couple of release?

Reply:
Sounds good to log first and raise after a couple of releases.

Reply:
I have opened #85863
I will mark this one as draft and rebase accordingly for `2023.4`